Added possibility to use `binary` field type for uuid in `DBALEventStore`

tPl0ch opened the pull request.

asm89 commented on src/Broadway/EventStore/DBALEventStore.php:
```diff
@@ -131,10 +140,25 @@ public function configureTable()
     {
         $schema = new Schema();
 
+        $uuidColumnDefinition = [
```
We still support php 5.3 (I know I know), can you use `array()` ?

asm89 commented on src/Broadway/EventStore/DBALEventStore.php:
```diff
@@ -170,4 +194,30 @@ private function deserializeEvent($row)
+        if ($this->useBinary) {
+            try {
+                return Uuid::fromString($id)->getBytes();
+            } catch (\Exception $e) {
```
Wondering if this should indeed be a no-op.

tPl0ch replied:
@asm89 Well, this is a kind of UUID specific feature, since the BINARY(16) definition matches exactly the uuid v4 byte length. We could also convert i.e. decimals (`dec2bin`) or other numeric values, but I don't see any benifit there.

asm89 replied:
Not handling the exception, but swallowing it leads to undefined behavior that might be hard to debug. I'm wondering if you should rethrow a proper exception.

tPl0ch replied:
@asm89 Well, we could force the usage of UUID for the binary mode and throw a `InvalidIdentifierException` or somthing along this line. What do you think?

asm89 replied:
@tPl0ch Yes, let's throw an exception if this fails.

asm89 commented:
Two comments, but overall I like this. :+1:

tPl0ch commented:
@asm89 I fixed the array issue, I added a new binary test id to an overridden data provider in the new TestCase. If you'd pass a binary UUID it's stored as binary, but you'd get a converted readable UUID string back when using `load($id)`.

asm89 commented on src/Broadway/EventStore/DBALEventStore.php:
```diff
@@ -170,4 +194,30 @@ private function deserializeEvent($row)
         );
     }
+
+    private function extractId($id)
```
`convertIdentifierToStorageValue()` ?

asm89 commented on src/Broadway/EventStore/DBALEventStore.php:
```diff
@@ -170,4 +194,30 @@ private function deserializeEvent($row)
+        return $id;
+    }
+
+    private function prepare($id)
```
`convertStorageValueToIdentifier()` ?

asm89 commented on test/Broadway/EventStore/EventStoreTest.php:
```diff
@@ -121,17 +122,28 @@ public function it_throws_an_exception_when_an_id_cannot_be_converted_to_a_strin
+                'Yolntbyaac'
             ),
-            array(
+            'String identitiy' => array(
```
nit: identity

asm89 commented on test/Broadway/EventStore/BinaryDBALEventStoreTest.php:
```diff
@@ -0,0 +1,87 @@
+            ),
+            'Binary UUID String' => array(
+                $uuid->getBytes(), // test UUID binary
+                $uuid->toString(),
```
I'm wondering if this test should be valid behavior? Isn't this confusing?

wjzijderveld commented on src/Broadway/EventStore/DBALEventStore.php:
```diff
@@ -45,12 +49,14 @@ public function __construct(
         $this->payloadSerializer  = $payloadSerializer;
         $this->metadataSerializer = $metadataSerializer;
         $this->tableName          = $tableName;
+        $this->useBinary          = Type::hasType('binary') && (bool) $useBinary;
```
Can we validate/assert this instead of silently calling back to the default? Maybe also validate it in the Extension?

wjzijderveld commented:
I like this as well :-)

tPl0ch commented:
ping @wjzijderveld @asm89 :)

wjzijderveld commented:
Thanks @tPl0ch  :+1:
